src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -1698,6 +1711,78 @@ private int GetTheBestCandidateIndex<TMember>(ArrayBuilder<MemberResolutionResul
         }
 
+        private static void RemoveLowerPriorityMembers<TMember>(ArrayBuilder<MemberResolutionResult<TMember>> results)
+            where TMember : Symbol
+        {
+            // - Then, the reduced set of candidate members is grouped by declaring type. Within each group:
+            //     - Candidate function members are ordered by *overload_resolution_priority*.
+            //     - All members that have a lower *overload_resolution_priority* than the highest found within its declaring type group are removed.
+            // - The reduced groups are then recombined into the final set of applicable candidate function members.
+
+            switch (results)
+            {
+                // Can't prune anything unless there's at least 2 candidates
+                case []:
+                case [_]:
+                // We only look at methods and indexers, so if this isn't one of those scenarios, we don't need to do anything.
+                case [{ Member: not (MethodSymbol or PropertySymbol { IsIndexer: true }) }, _, ..]:
+                    return;
+            }
+
+            // Attempt to avoid any allocations by starting with a quick pass through all results and seeing if any have non-default priority. If so, we'll do the full sort and filter.
+            if (results.All(r => r.LeastOverriddenMember.GetOverloadResolutionPriority() == 0))
+            {
+                // All default, nothing to do
+                return;
+            }
+
+            bool removedMembers = false;
+            var resultsByContainingType = PooledDictionary<NamedTypeSymbol, OneOrMany<MemberResolutionResult<TMember>>>.GetInstance();
+
+            foreach (var result in results)
+            {
+                var containingType = result.LeastOverriddenMember.ContainingType;
+                if (resultsByContainingType.TryGetValue(containingType, out var previousResults))
+                {
+                    var previousOverloadResolutionPriority = previousResults.First().LeastOverriddenMember.GetOverloadResolutionPriority();
+                    var currentOverloadResolutionPriority = result.LeastOverriddenMember.GetOverloadResolutionPriority();
+
+                    if (currentOverloadResolutionPriority > previousOverloadResolutionPriority)
+                    {
+                        removedMembers = true;
+                        resultsByContainingType[containingType] = OneOrMany.Create(result);
+                    }
+                    else if (currentOverloadResolutionPriority == previousOverloadResolutionPriority)
+                    {
+                        resultsByContainingType[containingType] = previousResults.Add(result);
+                    }
+                    else
+                    {
+                        removedMembers = true;
+                        Debug.Assert(previousResults.All(r => r.LeastOverriddenMember.GetOverloadResolutionPriority() > currentOverloadResolutionPriority));
```

Comment:
>> currentOverloadResolutionPriority [](http://example.com/codeflow?start=118&length=35)

Would this assert be stronger with `== previousOverloadResolutionPriorty` instead?